samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -323,7 +305,93 @@ private static void validateIntermediateStreamPartitions(JobGraph jobGraph) {
     }
   }
 
+  /**
+   * Ensures that all streams in supplied {@link JoinedStreamsGroup} agree in partition count.
+   * This includes setting partition counts of any intermediate streams that do not have their
+   * partition counts set.
+   */
+  private static void validateJoinedStreamsGroupPartitions(JoinedStreamsGroup joinedStreamsGroup) {
+    Set<StreamEdge> streamEdges = joinedStreamsGroup.getStreamEdges();
+    StreamEdge firstStreamEdgeWithSetPartitions =
+        streamEdges.stream()
+            .filter(streamEdge -> streamEdge.getPartitionCount() != StreamEdge.PARTITIONS_UNKNOWN)
+            .findFirst()
+            .orElse(null);
+
+    /*
+     * This group consists exclusively of intermediate streams with unknown partition counts.
+     * We cannot do any validation/computation of partition counts of such streams right here,
+     * but they are tackled later in the ExecutionPlanner.
+     */
+    if (firstStreamEdgeWithSetPartitions == null) {
+      return;
+    }
+
+    // Make sure all other stream edges in this group have the same partition count.
+    int partitions = firstStreamEdgeWithSetPartitions.getPartitionCount();
+    for (StreamEdge streamEdge : streamEdges) {
+      int streamPartitions = streamEdge.getPartitionCount();
+      if (streamPartitions == StreamEdge.PARTITIONS_UNKNOWN) {
+        streamEdge.setPartitionCount(partitions);
+        log.info("Inferred the partition count {} for the join operator {} from {}."
+            , new Object[] {partitions, joinedStreamsGroup.getGroupId(), firstStreamEdgeWithSetPartitions.getName()});
+      } else if (streamPartitions != partitions) {
+        throw  new SamzaException(String.format(
+            "Unable to resolve input partitions of stream %s for the join %s. Expected: %d, Actual: %d",
+            streamEdge.getName(), joinedStreamsGroup.getGroupId(), partitions, streamPartitions));
+      }
+    }
+  }
+
   /* package private */ static int maxPartitions(Collection<StreamEdge> edges) {
     return edges.stream().mapToInt(StreamEdge::getPartitionCount).max().orElse(StreamEdge.PARTITIONS_UNKNOWN);
   }
+
+  /**
+   * Represents a group of {@link StreamEdge}s participating in a Join operation.
+   */
+  /* package private */ static class JoinedStreamsGroup {
```

Comment:
`Joined` is an overloaded term and can mean multiple things.
Can we rename this to something else?

I am open to any suggestions here.

Reply:
I'm open to suggestions as well but let me ask you why the name felt confusing in this context. I thought it wasn't bad; a `JoinedStreamsGroup` is a group of `StreamEdge`s participating in a Join.

Reply:
Changed to `StreamEdgeSet`.